hf skills post with codex training llms (#3224)
huggingface/blog: hf-skills-training-codex.md

---
title: "Codex is Open Sourcing AI models"
thumbnail: /blog/assets/hf-skills-training/thumbnail-codex.png
authors:
- user: burtenshaw
- user: evalstate
---

# Codex is Open Sourcing AI models

![banner](https://huggingface.co/datasets/hf-skills/images/resolve/main/skills-codex-training-banner.png)

> Building on our work to get [Claude Code](https://huggingface.co/blog/hf-skills-training) to train open source models, we are now getting [Codex](https://developers.openai.com/codex/) to go further. We gave Codex access to the [Hugging Face Skills](https://github.com/huggingface/skills) repository. Which contains skills for Machine Learning and AI tasks like training or evaluating models. With HF skills, a coding agent can:

Reviewer: ```suggestion
Building on our work to get [Claude Code](https://huggingface.co/blog/hf-skills-training) to train open source models, we are now getting [Codex](https://developers.openai.com/codex/) to go further. We gave Codex access to the [Hugging Face Skills](https://github.com/huggingface/skills) repository, which contains skills for Machine Learning and AI tasks such as training or evaluating models. With HF skills, a coding agent can:
```

- Fine-tune and RL language models
- Review, explain, and act on training metrics in Trackio
- Evaluate checkpoints and act on evaluation results
- Create reports from experiments
- Quantize models to GGUF for local deployment
- Publish models to the Hub

This tutorial dives even deeper and shows you how it works and how to use it yourself. So let's get started. 

> [!NOTE]

> > Codex uses `AGENTS.md` files—packaged instructions, scripts, and domain knowledge—to accomplish specialized tasks. However, 'HF-skills' is completely agnostic to most major coding agents like Claude Code, Codex, or Gemini CLI. As well as any agent that uses `AGENTS.md`. 

Reviewer: "As well as..." -> Use it with any agent that uses `AGENTS.md` perhaps?

>
> We can install the skills directly from the repository. Codex will automatically detect the `AGENTS.md` file in the repository. For example:
>
> ```bash
> git clone https://github.com/huggingface/skills.git
> cd skills
> ```

With `HF-skills`, you can tell Codex something like:

```
Fine-tune Qwen3-0.6B on the dataset open-r1/codeforces-cots
```

And Codex will:

1. Validate your dataset format
2. Select appropriate hardware (t4-small for a 0.6B model)
3. Use and update a training script with Trackio monitoring
4. Submit the job to Hugging Face Jobs
5. Report the job ID and estimated cost
6. Check on progress when you ask
7. Help you debug if something goes wrong

The model trains on Hugging Face GPUs while you do other things. When it's done, your fine-tuned model appears on the Hub, ready to use.

> This isn't a toy demo. The extension supports the same training methods used in production: supervised fine-tuning, direct preference optimization, and reinforcement learning with verifiable rewards. You can train models from 0.5B to 70B parameters, convert them to GGUF for local deployment, and run multi-stage pipelines that combine different techniques.

Author: ```suggestion
This isn't a toy demo. The extension supports the same training methods used in production: supervised fine-tuning, direct preference optimization, and reinforcement learning with verifiable rewards. You can train models from 0.5B to 7B parameters, convert them to GGUF for local deployment, and run multi-stage pipelines that combine different techniques.
```

So let's get started!

## Setup and Install

Before starting, you'll need:

- A Hugging Face account with a [Pro](https://hf.co/pro) or [Team / Enterprise](https://hf.co/enterprise) plan (Jobs require a paid plan)
- A write-access token from [huggingface.co/settings/tokens](https://huggingface.co/settings/tokens)
- [Codex](https://developers.openai.com/codex/) installed and configured

### Install Codex

Codex is OpenAI's AI coding agent included in ChatGPT Plus, Pro, Business, Edu, and Enterprise plans. Codex brings AI assistance directly into your development workflow.

See the [Codex documentation](https://developers.openai.com/codex/) for installation and setup instructions.

> [!NOTE]
> Codex is available in ChatGPT Plus, Pro, Business, Edu, and Enterprise plans. It provides AI-powered code generation and assistance directly in your IDE or terminal.

> 

Reviewer: ```suggestion
```

Already mentioned above.

```bash
hf auth login
```

> [!TIP]
> Codex supports [MCP (Model Context Protocol)](https://developers.openai.com/codex/) servers. You can configure the Hugging Face MCP server for additional Hub integration capabilities. You can add the Hugging Face MCP server to your Codex configuration by adding the following to your `~/.codex/config.toml` file:
>
> ```toml
> [mcp_servers.huggingface]
> command = "npx"
> args = ["-y", "mcp-remote", "https://huggingface.co/mcp?login"]
> ```
>
> Configure Hugging Face MCP Server to use relevant MCP servers like Jobs in the [Settings](https://huggingface.co/settings/mcp) page.

Then start Codex and you'll be directed to the Hugging Face MCP authentication page.

## Your first AI Experiment

Let's walk through a complete example. We'll fine-tune a small model to improve code solving abilities, using the open-r1/codeforces-cots dataset and the openai_humaneval benchmark.

### Instruct Codex to do an end-to-end fine-tuning experiment

Start Codex in your project directory. Then give it a simple and clear instruction:

```
Start a new fine-tuning experiment to improve code solving abilities on using SFT. 
- Maintain a report for the experiment. 

> - Evaluate models the openai_humaneval benchmark

Reviewer 2: very cool that you can tell it's a benchmark without giving dataset id directly

> - Evaluate models the openai_humaneval benchmark

Reviewer: ```suggestion
- Evaluate models with the openai_humaneval benchmark
```

| Weight Decay | 0.01 |
| Warmup Ratio | 0.03 |
| Eval Strategy | steps (500) |
| Save Strategy | steps (500), `hub_strategy=every_save`, limit=2 |
| Precision | bf16 |
| Gradient Checkpointing | true |
| Packing | false |
| Hub Model | `burtenshaw/qwen3-codeforces-cots-sft` |
| Hardware | a10g-small |
| Timeout | 2h |
| Trackio | project `qwen3-codeforces-cots`, run `sft-a10g` |

## Run Status
In Progress (queued to submit)

## Run Logs
Pending submission (job link will be added)

## Trackio Logs
Pending (will link after job starts)

## Run Evaluations
Pending (lighteval `openai_humaneval` for base + checkpoints)

---

# Experiment Evaluations
| Run Title | Benchmark | Score | Evaluation Job Link | Model Link |
|-----------|-----------|-------|---------------------|------------|
| `sft-a10g` - `TBD` - `In Progress` | HumanEval pass@1 | TBD | TBD | [burtenshaw/qwen3-codeforces-cots-sft](https://huggingface.co/burtenshaw/qwen3-codeforces-cots-sft)

```

</details>

>[!NOTE]
> The `open-r1/codeforces-cots` dataset is a dataset of codeforces problems and solutions. It is a good dataset for instruction tuning a model to solve hard coding problems.

### Updating the Training Report

> As the experiment progresses, Codex will update the report with the latest information and each run report. You can update the report by editing the `training_reports/<model>-<dataset>-<method>.md` file.

Author: No. They agent maintains the progress report based on what of its skills.
